ENH: OpenVMS port

vorfol opened the pull request.

seberg committed:
BUG: Remove temporary change of descr/flags in VOID functions

This is done using a hack around stack allocated arrays, which
requires some dark magic with respect to the setting the base
object (we skip normal arrays and INCREF/DECREF during base setting,
which _does_ work for such static arrays, but only if we give them
a (not quite) fake refcount of 1 and set the base and their type
correctly to PyArrayType.

Closes gh-15387

seberg committed:
TST: Add test for nonzero and copyswapn (through advanced indexing)

seberg committed:
BUG: Ignore fewer errors during array-coercion

This changes it so that we only ignore attribute errors on
looking up `__array__` and propagate errors when checking
for sequences `len(obj)` if those errors are either
RecursionError or MemoryError (we consider them unrecoverable).

Also adds test for bad recursive array-like with sequence
as reported in gh-17785. The test might be flaky/more complicated
in which case it should probably just be deleted.

seberg committed:
ENH: Micro-optimize where=True path for mean, var, any, and all

This removes a 20%-30% overhead, and thus the largest chunk of
slowdown incurred by adding the `where` argument. Most other places
have fast-paths for `where=True`, this one also should have it.

The additional argument does slow down the function versions a bit
more than this, but that is to be expected probably (it has to
build a new argument dict, at some point we might want to move this
to C, but that seems worth much more with FASTCALL logic).

seberg committed:
DOC: Add where to all, any, mean, var, and std method docstrings

t-karatsu committed:
ENH: add support for fujitsu compiler to numpy. (#17792)

charris committed:
DOC: Update changelog and notes for 1.20.0

charris committed:
Merge pull request #17904 from charris/backport-17896

ENH: Speed up default `where` in the reduce-like method

charris committed:
REL: NumPy 1.20.0rc1 release.

fxcoudert committed:
Update gnu.py

charris committed:
REL: prepare 1.20.x for further development

cgohlke committed:
Fix AttributeError: 'bool' object has no attribute 'ndim'

vorfol committed:
fix stdint

dhomeier committed:
TST: remove stray '+' from f-string upgrade

vorfol committed:
merge try one

vorfol committed:
define #ctype

vorfol committed:
back to before merge

seberg committed:
DOC: Update release notes to mention `type(dtype) is not np.dtype`

This also requires mentioning the C-API macro
`#define PyArray_DescrCheck(op) PyObject_TypeCheck(op, &PyArrayDescr_Type)`

which was updated to the above in 1.16.6 meaning that using the
macro and compiling against an older NumPy version will cause
issues.  The macro has to be avoided in that case.

[ci skip]

vorfol committed:
merge with github

charris committed:
Merge pull request #17916 from charris/backport-17792

ENH: add support for fujitsu compiler to numpy.

charris committed:
Merge pull request #17915 from charris/backport-17912

TST: remove stray '+' from f-string upgrade

charris committed:
Merge pull request #17922 from charris/backport-17907

BUG: 'bool' object has no attribute 'ndim'

vorfol committed:
fix OpenVMS issues (abs)

vorfol committed:
Merge branch 'master' into master

vorfol committed:
use NPY_SIZEOF_PY_INTPTR_T

vorfol committed:
cast from char** to intptr_t

charris committed:
Merge pull request #17931 from charris/backport-17914

DOC: Update release notes to mention `type(dtype) is not np.dtype`

vorfol committed:
warning was print intptr_t as %d

vorfol committed:
get rid of stdint.h

vorfol committed:
skip test based by Parser

vorfol committed:
correct comments

eric-wieser commented:
Does the OpenVMS compiler claim to support C99?

Roughly speaking, there seem to be two things  that this PR is trying to fix.

* Numpy makes assumptions about `size_t` vs `intptr_t` which the C standard does not require to be true. Fixing these is probably a good idea (but a lot of work), if we can do so without breaking the ABI.
* The OpenVMS compiler is not standard-compliant:
  * longdoubles treat `nan` incorrectly (IEEE-754 maybe applies?).
  * It reserves identifiers that are not labelled as reserved by the C standard.
 I'm wary of carrying around the maintenance burden for both a niche and non-conforming compiler.

eric-wieser commented on numpy/core/memmap.py:
```diff
@@ -253,6 +254,7 @@ def __new__(subtype, filename, dtype=uint8, mode='r+', offset=0,
                 fid.seek(bytes - 1, 0)
                 fid.write(b'\0')
                 fid.flush()
+                fsync(fid.fileno())
```
What's the motivation for this change? Is it an improvement on other platforms too?

vorfol replied:
Maybe other platforms do synchronization inside flush(). But it is not obligatory. Without fsync() the next call of mmap() might fail. On OpenVMS it fails.

mattip replied:
See point 9.3 in the [openvms faq](http://www.hoffmanlabs.org/vmsfaq/vmsfaq_013.html). Although that seems like an anti-feature, I would expect [`fflush`](https://linux.die.net/man/3/fflush) to be enough.

eric-wieser commented on numpy/core/src/multiarray/nditer_constr.c:
```diff
@@ -2065,7 +2065,8 @@ npyiter_compute_index_strides(NpyIter *iter, npy_uint32 flags)
         if (itflags & NPY_ITFLAG_HASINDEX) {
             axisdata = NIT_AXISDATA(iter);
-            NAD_PTRS(axisdata)[nop] = 0;
+            // reset the index, it is just beyond the pointers
+            *(npy_intp*)&(NAD_PTRS(axisdata)[nop]) = 0;
```
I think an `NAD_INDEX` macro might be good to encapsulate this cast.

eric-wieser commented on numpy/random/tests/test_extending.py:
```diff
@@ -57,9 +58,9 @@ def test_cython(tmp_path):
     with open(build_dir / 'extending.c') as fid:
-        txt_to_find = 'NumPy API declarations from "numpy/__init__.pxd"'
+        txt_to_find = re.compile(r'NumPy API declarations from "numpy\/__init__(\.([^.]*?))?.pxd"')
         for i, line in enumerate(fid):
-            if txt_to_find in line:
+            if txt_to_find.search(line):
                 break
```
This looks like a bugfix for the new cython change, right? Perhaps we should create a new PR with this.

rossbar committed:
TST: Fix crosstalk issues with polynomial str tests.

Polynomial printing tests implicitly depended on calling
order, causing the test suite to fail when the test ordering was
randomized with the pytest-random plugin (gh-17954).

Two factors contributed to this:
 * Improper setting of class-level test config and
 * Poorly designed test that overrode an inherited class
   variable.

seberg committed:
TST: Ensure `like=` tests are not sensitive to execution order

This moves class creation into a `setup` function and uses `self`
instead of the test class all over.  There are probably nicer
ways to fix (and improve) it, but this seemed nice minimal.

seberg committed:
TST: Do not disable errors globally in ma/timer_comparison.py

This script is meant to be run stand-alone (and should probably
just not be at this place), but moving these down is cleaner in
any case.

The issue is that our "public API" tests import the module and
then modify the global state.

seberg committed:
DEP: Futurewarn on requiring __len__ on array-likes

This fixes issue gh-17965.  The slightly annoying thing is that
there is no simple way to opt-in to the new behaviour and the old
behaviour is a bit quirky to begin with (honoring the dtype, but
not the shape).

charris committed:
BUG: Replace f-string in setup.py

We want the warning to stay compatible with early versions of python.

seberg committed:
Merge pull request #17990 from charris/fix-f-string-in-setup

BUG: Replace f-string in setup.py

seiko2plus committed:
BLD, BUG: Fix detecting aarch64 on macOS

Co-authored-by: h-vetinari <h.vetinari@gmx.com>

mattip committed:
BLD: update to OpenBLAS 0.3.13

charris committed:
Merge pull request #18019 from charris/backport-18011

BLD: update to OpenBLAS 0.3.13

charris committed:
Merge pull request #18018 from charris/backport-17962

TST: Ensure tests are not sensitive to execution order

charris committed:
Merge pull request #18017 from charris/backport-17957

TST: Fix crosstalk issues with polynomial str tests.

charris committed:
Merge pull request #18016 from charris/backport-17906

BUG: Fix a MacOS build failure

charris committed:
Merge pull request #18015 from charris/backport-17817

BUG: Ignore fewer errors during array-coercion

charris committed:
Merge pull request #18024 from charris/backport-17973

DEP: Futurewarn on requiring __len__ on array-likes

r-devulap committed:
BUG: make a variable volatile to work around clang compiler bug (#18030)

* BUG: make a variable volatile to work around clang compiler bug

* Adding comments for relevance

* TST: Adding test to check for no overflow warnings in log

Fixes #18005

charris committed:
Merge pull request #18035 from charris/backport-18030

BUG: make a variable volatile to work around clang compiler bug

seiko2plus committed:
BLD, MAINT: add platform info to the final optimization report

mattip committed:
TST: add back sdist test run

charris committed:
Merge pull request #18049 from charris/backport-18046

TST: add back sdist test run

seberg committed:
BUG: Fix concatenation when the output is "S" or "U"

Previously, the dtype was used, this now assumes that we want to
cast to a string of (unknown) length.  This is a simplified version
of what happens in `np.array()` or `arr.astype()` (it does never
inspect the values, e.g. for object casts).

This is more complex as I would like, and with the refactor of
ResultType and similar can be cleaned up a bit more hopefully.

Note that currently, object to "S" or "U" casts simply return
length 64 strings, but with the new version, this will be an error
(although the error message probably needs improvement).
This is a behaviour inherited from other places however.

charris committed:
Merge pull request #18063 from charris/backport-18052

BUG: Fix concatenation when the output is "S" or "U"

charris committed:
Merge pull request #18064 from charris/backport-18001

BLD, BUG: Fix detecting aarch64 on macOS

charris committed:
REL: Prepare for 1.20.0rc2 release.

- Update .mailmap
- Update 1.20.0-changelog.rst
- Update 1.20.0-notes.rst

charris committed:
Merge pull request #18068 from charris/prepare-1.20.0rc2-release

REL: Prepare for 1.20.0rc2 release.

charris committed:
REL: NumPy 1.20.0rc2 release.

charris committed:
REL: prepare 1.20.x for further development

tylerjereddy committed:
MAINT: crackfortran regex simplify

* remove extraneous character class markers used in
`crackline_re_1`: `\w` and `=` on their own have no
benefit to character class `[]` inclusion

* `name_match` has a character class that can be
simplified because `\w` metacharacter already
encompasses the digit metacharacter and the
underscore

tylerjereddy committed:
MAINT: PR 18072 revisions

* `name_match` regular expression now starts by
matching a letter only, based on reviewer
feedback

seiko2plus committed:
BUG, SIMD: Fix _simd module build for 64bit ARM/NEON clang

seiko2plus committed:
BUG, BLD: Generate the main dispatcher config header into the build dir

  The new path becomes `build/src.*/numpy/distutils/include/npy_cpu_dispatch_config.h`
  instead of `numpy/core/src/common/_cpu_dispatch.h`.

  The new path allows other projects to re-use the CPU dispatcher
  once we decide to expose the following headers:
    - `numpy/core/src/common/npy_cpu_dispatch.h`
    - `numpy/core/src/common/npy_cpu_features.h`

seiko2plus committed:
MAINT, BLD: few tweaks in the comments and log message

Co-authored-by: Matti Picus <matti.picus@gmail.com>
Co-authored-by: h-vetinari <h.vetinari@gmx.com>
Co-authored-by: Derek Homeier <dhomeie@gwdg.de>

charris committed:
Merge pull request #18108 from charris/backport-18100

BUG, BLD: Generate the main dispatcher config header into the build dir

seberg committed:
BUG: Fix promotion of half and string

I somehow managed to miss that half breaks the order of dtypes and
has a higher number than the strings. Could be backported, but it
doesn't really matter, since it only makes a difference if the
compile time flag is used and even then is pretty fringe.

seiko2plus committed:
BLD: Add extra check for AVX512DQ mask conversions

seiko2plus committed:
BUG, MAINT: improve avx512 mask logical operations

  It also fixes conversion warning between `__mmask16` and `__mmask8`
  on msvc2019 when logical intrinsics of AVX512DQ are available.

charris committed:
Merge pull request #18120 from charris/backport-18086

BUG, SIMD: Fix _simd module build for 64bit ARM/NEON clang

charris committed:
REL: Update 1.20.x after 1.19.5 release.

charris committed:
Merge pull request #18127 from charris/backport-18126

REL: Update 1.20.x after 1.19.5 release.

charris committed:
Merge pull request #18130 from charris/backport-18115

BUG: Fix promotion of half and string

charris committed:
Merge pull request #18146 from charris/backport-18118

BUG, MAINT: improve avx512 mask logical operations

seberg committed:
BUG: Ensure too many advanced indices raises an exception

The number of indices is limited to 2*MAXDIMS currently to allow
mixing integer indices, e.g. with new indices `np.newaxis` (one
removes output dimensions, the other adds new ones).
This means that more than MAXDIMS advanced indices can be passed
on to the advanced indexing machinery (`MapIterNew`), which did
not check for this possibility.

Closes gh-18145

seberg committed:
BUG: Promotion between strings and objects was assymetric

After an unrelated fix, the logic for string and object promotion
was incorrect briefly, this fixes it to be correct (symmetric).
Before, string and unicode would return that
`string.__common_dtype(object)` is in fact `string`, which is of
course incorrect. (since `object.__common_dtype__(other)` always
returns `object` this depended on the order, and the NumPy tests
apparently did only test the opposite direction (or nothing).

charris committed:
Merge pull request #18154 from charris/backport-18152

BUG: Promotion between strings and objects was assymetric

pearu committed:
BUG: Fix detecting Fortran 90+ source format.

pearu committed:
BUG: subroutines using assumed shape arrays must be treated as F90 codes. Don't split inline comments.

pearu committed:
BUG: f2py specific __user__ modules should not be used in Fortran sources to be compiled. Closes #17797.

pearu committed:
Add test for gh17797.

pearu committed:
Apply reviewers comments.

pearu committed:
Apply reviewer comments.

Commit:
STY: Use explicit reexports for numpy.typing objects

`mypy --strict` is disabling `implicit_reexport`.  Consequently, when we try to
 import `ArrayLike` and `DTypeLike` from `numpy.typing`, mypy throws an error.

With this commit we add explicit "reexports" for these two objects.

Co-Authored-By: pmav99 <411196+pmav99@users.noreply.github.com>

charris committed:
Merge pull request #18192 from BvB93/fix_18190

MAINT: Use explicit reexports for numpy.typing objects

pearu committed:
BUG: threads.h existence test requires GLIBC > 2.12. (#18180)

* defined(__STDC_NO_THREADS__) can be trusted only when using glibc > 2.12

* Add __GLIBC__ defined check.

pearu committed:
Applied reviewers comments: add ident

mattip committed:
BLD: update OpenBLAS to af2b0d02

seberg committed:
BUG: Keep ignoring most errors during array-protocol lookup

Closes (the later point) in gh-17965 and reverts parts of
gh-17817.
Shapely did rely on being able to raise a NotImplementedError
which then got ignored in the attribute lookup.
Arguably, this should probably just raise an AttributeError to
achieve that behaviour, but it means we can't just rip the band-aid
off here.

Since 1.20 is practically released, just reverting most of the change
(leaving only recursion and memory error which are both arguably
pretty fatal).
Ignoring most errors should be deprecated (and I am happy to do so),
but it is not important enough for 1.20 or very important in itself.

Closes gh-17965

charris committed:
Merge pull request #18201 from charris/backport-18197

BUG: Keep ignoring most errors during array-protocol lookup

rgommers committed:
MAINT: warn when shuffling unrecognized objects

Closes gh-18206

rgommers committed:
BUG: shuffling empty array with axis=1 was broken

This would trigger:
```
NotImplementedError: Axis argument is only supported on ndarray objects
```
because empty arrays and array scalars would take the "untyped" path.

The bug exists only for Generator, not in RandomState (it doesn't have
axis keyword for `shuffle`), but update both because it keeps
implementations in sync and the change results in more understandable
code also for RandomState.

rgommers committed:
Remove duplicate axis check

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

charris committed:
Merge pull request #18219 from charris/backport-18211

MAINT: random shuffle: warn on unrecognized objects, fix empty array bug

vorfol committed:
version is released

vorfol committed:
remove dep ver

Commit:
MAINT: Remove `__all__` in favor of explicit reexports

seberg committed:
DOC: Clarify the type alias deprecation message

This tries to clarify the type alias deprecation message slightly
to give more guidance on how to review the replacement (if desired).

charris committed:
Merge pull request #18231 from h-vetinari/1.20.x

BLD: update OpenBLAS to af2b0d02

seberg committed:
Update doc/source/release/1.20.0-notes.rst

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

seberg committed:
Update numpy/__init__.py

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

seberg committed:
Apply suggestions from code review

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

rossbar committed:
Fix ref to user guide in release notes.

charris committed:
Merge pull request #18237 from charris/backport-18230

DOC: Clarify the type alias deprecation message

mattip committed:
add an 'apt update'

seberg committed:
Update numpy/core/src/multiarray/mapping.c

charris committed:
DOC: Prepare for the NumPy 1.20.0 release.

- Update 1.20.0-notes.rst
- Update 1.20.0-changelog.rst

[ci skip]

charris committed:
Merge pull request #18258 from charris/backport-18255

MAINT: add an 'apt update'

charris committed:
Merge pull request #18257 from charris/backport-18150

BUG: Ensure too many advanced indices raises an exception

charris committed:
Merge pull request #18259 from charris/prepare-for-1.20.0-release

DOC: Prepare for the NumPy 1.20.0 release.

charris committed:
REL: NumPy 1.20.0 release.

charris committed:
TST: Don't fail check_large_zip on signal 9

If the test is terminated for lack of memory the intended zip file check
does not take place, raise xfail instead.

charris committed:
REL: NumPy 1.20.0 release with fix.

charris committed:
REL: prepare 1.20.x for further development

mckib2 committed:
FIX: random: C linkage for random distributions

mckib2 committed:
BUG: cffi parser ignores __cplusplus guards

tylerjereddy committed:
MAINT: gracefully shuffle memoryviews

* allow graceful shuffling of memoryviews, with
same behavior as arrays, instead of producing
a warning on `memoryview` shuffle

charris committed:
BUG: Fix typo in ``numpy.__init__.py``

Fix use of "long" when "unicode" was intended. Also fix
spelling of "unicode".

Closes #18287

tylerjereddy committed:
MAINT: PR 18282 revisions

* based on reviewer feedback, instead of explicitly
supporting `memoryview` shuffling via `asarray()`, support
the shuffling implicitly by using `Sequence` instead
of `MutableSequence`

Commit:
ENH: Add annotations for certain module-level dunders

Backport of https://github.com/numpy/numpy/pull/18229/commits/bcfb1180374a50c34e11e576efe00d5c676d9f77

Commit:
MAINT: Add missing `__all__` annotations

Backport of https://github.com/numpy/numpy/pull/18302/commits/12d99670040c265a615694fd3b8ae1492807c93f

Commit:
MAINT: Add previously-missing placeholder annotations

Backport of https://github.com/numpy/numpy/pull/18302/commits/c21509a595eea2cf9d64893a58adb5abcce68bb1

charris committed:
Merge pull request #18306 from BvB93/missing-attr2

MAINT: Add missing placeholder annotations

charris committed:
Merge pull request #18310 from charris/backport-18293

BUG: Fix typo in ``numpy.__init__.py``

mckib2 committed:
BUG: don't mutate list of fake libraries while iterating over it (#18295)

* BUG: don't mutate list of fake libraries while iterating over it

* BUG: iterate over copy of list

* TST: add build test for build_ext fix (#1)

* TST: add build test for build_ext fix

* TST: clearer test name

* STY: use triple quotes instead of lists of strings

* FIX: check for f77 compiler before test is run

* DOC: add comment explaining that a list copy is necessary

seberg committed:
BUG: Allow unmodified use of isclose, allclose, etc. with timedelta

Disallowing timedelta64+float promotion (to timedelta64) in all cases
(previously it was assymetric and "half allowed") meant that isclose,
allclose, np.ma.allclose, and assert_arrays_almost_equal (which uses
isclose), would stop work for timedelta64. Hardcoding that timedelta64
is passed on unmodified retains the old behaviour.

It may make sense to deprecate or change this behaviour in the future,
but for the 1.20 release, the behaviour should be as much unmodified
as possible.

Closes gh-18286

charris committed:
Merge pull request #18326 from charris/backport-18295

BUG: don't mutate list of fake libraries while iterating over it.

charris committed:
Merge pull request #18327 from charris/backport-18282

MAINT: gracefully shuffle memoryviews

charris committed:
Merge pull request #18328 from charris/backport-18264

BUG: Use C linkage for random distributions

seberg committed:
BUG: Allow pickling all relevant DType types/classes

Introducing the metaclass without a canonical top-level name broke
pickling of `type(np.dtype(...))` (which was always just `np.dtype`).

While a better solution will likely be possible by making the
DTypes HeapTypes and this solution may not work for all imaginable
cases (i.e. it is plausible for a dtype to not have a scalar type
associated), using a `copyreg` registration for the metaclass
surprisingly works without any issues and seems like the simplest
solution right now.

Closes gh-16692, gh-18325

rgommers committed:
CI: fix when GitHub Actions builds trigger, and allow ci skips

Similar to what SciPy does. Right now, it triggers even on pushes
to branches on forks, which is useless and generates lots of
notifications on failures and wasted resources if no failures.

charris committed:
Merge pull request #18336 from charris/backport-18333

CI: fix when GitHub Actions builds trigger, and allow ci skips

charris committed:
REL: Prepare for the NumPy 1.20.1 release.

- Create 1.20.1-changelog.rst
- Create 1.20.1-notes.rst
- Update .mailmap

charris committed:
Merge pull request #18337 from charris/backport-18329

BUG: Allow unmodified use of isclose, allclose, etc. with timedelta

charris committed:
Merge pull request #18345 from charris/backport-18332

BUG: Allow pickling all relevant DType types/classes

seberg committed:
CI: CircleCI seems to occasionally time out, increase the limit

CircleCI seems to default to a 10 minute timeout on no output,
simply increase it to 30 minutes to ensure the build should
succeed.
It seems the build does take 25+ minutes (maybe depending on the
machine it runs on). Technically, 30 minutes is probably more than
necessary, since there is a print in the middle, but that is
just a warning that could go away at some point.

pearu committed:
BUG: Fix missing signed_char dependency. Closes #18335.

dmontesl committed:
DOC: Change license date 2020 -> 2021 (gh-18346)

A very small contribution to update the license date from 2020 to 2021 :)

charris committed:
Merge pull request #18352 from charris/backport-18346

 DOC: Change license date 2020 -> 2021

charris committed:
Merge pull request #18353 from charris/backport-18349

CI: CircleCI seems to occasionally time out, increase the limit

charris committed:
Merge pull request #18354 from charris/backport-18184

BUG: Fix f2py bugs when wrapping F90 subroutines.

charris committed:
Merge pull request #18357 from charris/backport-18180

BUG: threads.h existence test requires GLIBC > 2.12.

charris committed:
Merge pull request #18356 from charris/backport-18072

MAINT: crackfortran regex simplify

charris committed:
Merge pull request #18351 from charris/backport-18350

BUG: Fix missing signed_char dependency. Closes #18335.

charris committed:
Merge pull request #18359 from charris/prepare-for-1.20.1-release

REL: Prepare for the NumPy 1.20.1 release.

charris committed:
REL: NumPy 1.20.1 release.

charris committed:
REL: prepare 1.20.x for further development

charris committed:
MAINT: Update f2py from master.

This will make it easier to backport the f2py fixes currently going
into master.

charris committed:
Merge pull request #18382 from charris/update-1.20.x-f2py

MAINT: Update f2py from master.

seberg committed:
BUG: Fix tiny memory leaks when `like=` overrides are used

I thought I had fixed these leaks, but it appears I missed some.
We probably should backport this to 1.20.x (its simple), but the leaks
are also pretty harmless unless someone uses `like=` hundrets of
thousands of times in a running program (and its a new fetaure).

seberg committed:
BUG: Fix refcount leak in f2py `complex_double_from_pyobj`

ahaldane committed:
BUG: Segfault in nditer buffer dealloc for Object arrays

Commit:
BUG: Fixed an issue where `diagflat` could overflow on windows and 32-bit platforms

charris committed:
Merge pull request #18459 from charris/backport-18455

BUG: ``diagflat`` could overflow on windows or 32-bit platforms

charris committed:
Merge pull request #18460 from charris/backport-18436

BUG: Fix refcount leak in f2py ``complex_double_from_pyobj``.

charris committed:
Merge pull request #18461 from charris/backport-18420

BUG: Fix tiny memory leaks when ``like=`` overrides are used

charris committed:
Merge pull request #18462 from charris/backport-15392

BUG: Remove temporary change of descr/flags in VOID functions

charris committed:
Merge pull request #18469 from charris/backport-18450

BUG: Segfault in nditer buffer dealloc for Object arrays

cgohlke committed:
MAINT: Remove suspicious type casting

cgohlke committed:
MAINT: remove nonsensical comparison of pointer < 0

cgohlke committed:
MAINT: verify pointer against NULL before using it

cgohlke committed:
BUG: check if PyArray_malloc succeeded

cgohlke committed:
MAINT: cast Py_ssize_t to int

cgohlke committed:
Goto done when an error condition is reached

charris committed:
Merge pull request #18485 from charris/backport-18476

BUG: Remove suspicious type casting

charris committed:
Merge pull request #18486 from charris/backport-18477

BUG: remove nonsensical comparison of pointer < 0

charris committed:
Merge pull request #18487 from charris/backport-18478

BUG: verify pointer against NULL before using it

charris committed:
Merge pull request #18488 from charris/backport-18479

BUG: check if PyArray_malloc succeeded

ahaldane committed:
BUG: incorrect error fallthrough in nditer

Fixup to gh-18450

charris committed:
CI: Use Ubuntu 18.04 to run "full" test.

NumPy does not build using the `--coverage` flag on Ubuntu 20.04, the
problem seems to be gcc 9.3.0-17. Work around that by running on Ubuntu
18.04 instead.

ExpHP committed:
BUG: NameError in numpy.distutils.fcompiler.compaq

Fix a simple mistake in commit da0497fdf35 which can produce a NameError when installing numpy in MinGW/MSYS2.

charris committed:
CI: Pin docker image.

The azure-pipeline test Linux_Python_38_32bit_full_with_asserts has
been failing after the release of the latest manylinux2010. Pin
the docker image to an earlier version to fix this.

Closes #18553.

charris committed:
Merge pull request #18559 from charris/backport-ci-fixes

CI: Backport CI fixes from main.

charris committed:
Merge pull request #18546 from ahaldane/backport_18530

BUG: incorrect error fallthrough in nditer

mtsokol committed:
BUG: Fixed ``where`` keyword for ``np.mean`` & ``np.var`` methods (gh-18560)

* Fixed  keyword bug

* Added test case

* Reverted to original notation

* Added tests for var and std

Closes gh-18552

seberg committed:
BUG: Fix ma coercion list-of-ma-arrays if they do not cast to bool

There was a regression here due to force casting to bool, but if that
happens to fail (it does, but should not for strings). The mask
would just be dropped.

Of course masked arrays are held together by faith here, but its
a regression.

Closes gh-18551

seberg committed:
Apply suggestions from code review

BvB93 committed:
MAINT: Add annotations for `dtype.__getitem__`, `__mul__` and `names` (#18599)

seberg committed:
Update numpy/ma/tests/test_core.py

charris committed:
Merge pull request #18612 from charris/backport-18560

BUG: Fixed ``where`` keyword for ``np.mean`` & ``np.var`` methods

charris committed:
Merge pull request #18611 from charris/backport-18539

BUG: NameError in numpy.distutils.fcompiler.compaq

charris committed:
CI: Update apt package list before Python install

Closes #18613.

Commit:
API: Formally classify `np.lib.stride_tricks` as part of the public API

With `as_strided`, and the newly introduced `sliding_window_view` function, there are currently 2 public objects that can:
a. Only be imported from a private module
b. Are publicly documented to-be imported from aforementioned module

Both observations are problematic and in need of rectification.
This commit therefore moves `np.lib.stride_tricks` to the `PUBLIC_MODULES` list.

Commit:
MAINT: Re-export a number of sub-modules

Ensures that type checkers will allow the likes of:

>>> import numpy as np
>>> out = np.lib.stride_tricks.sliding_window_view(...)

charris committed:
Merge pull request #18617 from charris/fix-gcc48-test

CI: Update apt package list before Python install

Commit:
API: Move `polynomial.polyutils` to the `PRIVATE_BUT_PRESENT_MODULES` list

Aforementioned module was accidently marked as public

charris committed:
Merge pull request #18636 from BvB93/modules2

MAINT: Ensure that re-exported sub-modules are properly annotated

charris committed:
Merge pull request #18638 from charris/backport-18605

BUG: Fix ma coercion list-of-ma-arrays if they do not cast to bool

seberg committed:
BUG: Fix small valgrind-found issues (#18651)

* BUG: Fix small valgrind-found issues

This should be backportable. There was at least one that I could
not reproduce when running the tests again. And the new random-shuffle
tests give false-positives (which is just slightly annoying, considering
that we are very close to almost only "longdouble" related
false-positives)

* BUG: Add missing decref in user-dtype fallback paths

The missing decref here only leaks references and can never leak
actual memory fortunately.

* MAINT,TST: Simplify the "refcount logic" in the dispatch tests again

Using SETREF can be nice, but was just overcomplicating thing here...

charris committed:
Merge pull request #18661 from charris/backport-18651

BUG: Fix small valgrind-found issues (#18651)

charris committed:
REL: Prepare for the NumPy 1.20.2 release.

- Create the 1.20.2-changelog
- Update the 1.20.2-notes

seberg committed:
BUG: Fix small issues found with pytest-leaks

None of these are particularly worrying as they either usually only
leak reference (and not memory) or appear in rare or almost impossible
error-paths, or are limited to the tests.

Unfortunately, this PR will not apply to 1.20.x, due to small changes
in the overrides.

Backport of gh-18670 (with two small differences)

charris committed:
Merge pull request #18671 from seberg/backport-small-pytest-leaks-fixes

BUG: Fix small issues found with pytest-leaks

charris committed:
Merge pull request #18681 from charris/prepare-1.20.2-release

REL: Prepare for the NumPy 1.20.2 release.

charris committed:
REL: NumPy 1.20.2 release.

charris committed:
REL: prepare 1.20.x for further development

aarchiba committed:
BUG: fix segfault in object/longdouble operations

The operation None*np.longdouble(3) was causing infinite recursion as it
searched for the appropriate conversion method. This resolves that, both
for general operations and for remainders specifically (they fail in a
subtly different way).

Closes #18548

aarchiba committed:
Remove unused variables

kchoudhu committed:
Strip extra newline when dumping gfortran version on MacOS

charris committed:
MAINT: Allow more recursion depth for scalar tests.

The test_operator_object_left and test_operator_object_right tests in
test_scalarmath.py set a maximum recursion depth of 100 that caused an
error in gitpod because it was too close to the starting depth of 75.
The fix here is to raise the maximum depth to 200.

Closes #18749

vorfol committed:
back readonly

vorfol committed:
back _align

vorfol committed:
%z as %PY_FORMAT_SIZE_T

vorfol committed:
get rid of extra compiler args

Commit:
BUG: Correct `datetime64` missing type overload for `datetime.date`

vorfol committed:
merged with 1.20.2

charris committed:
Merge pull request #18764 from BvB93/re-export

MAINT: Remove `__all__` in favor of explicit re-exports

charris committed:
Merge pull request #18763 from BvB93/datetime

BUG: Correct `datetime64` missing type overload for `datetime.date` #18640

charris committed:
Merge pull request #18768 from charris/backport-18701

BLD: Strip extra newline when dumping gfortran version on MacOS

charris committed:
Merge pull request #18769 from charris/backport-18691

BUG: fix segfault in object/longdouble operations

Commit:
MAINT: Ensure that the `np.ndarray` constructor can take any shape-like oject

Commit:
MAINT: Relax the integer-type-constraint of `npt._ShapeLike`

Change `int` into the more forgiving `SupportsIndex` protocol

Commit:
STY: pep8 fixes

Commit:
TST: Fix a test failure for python 3.7

seberg committed:
CI: Use `towncrier build` explicitly

Without any argument towncrier defaults to `build`.  But towncrier
added a `--version` argument to just print the towncrier version.

NumPy relies on passing `--version` which collides.  Even if
towncrier might revert the change, using `towncrier build` explicitly
just doesn't hurt.

Closes gh-18788

charris committed:
MAINT: Use recent towncrier releases on PyPI.

We were installing towncrier from the github master branch because there
had been no releases with the needed fixes. That has changed with the
recent release of versions 19.9.0 and 21.3.0.

charris committed:
Merge pull request #18794 from charris/backport-18790

MAINT: Use towncrier build explicitly

Commit:
MAINT: Relax the miscellaneous integer-type-constraints

seberg committed:
BUG: Initialize the full nditer buffer in case of error

This is necessary because in some rare cases (reductions), we may
not actually use the full buffer.  In that case, the cleanup-on-error
code would have to grow smart enough to handle these cases.
It seems much simpler to just always initialize the full buffers, even
if we may not end up using them.

Admittedly, the old logic might have skipped the buffer clearing
(especially the full buffer) in a fair bit of cases, but since
this is only relevant for `object` dtype, I assume this is fine.

seberg committed:
MAINT: Remove buffer-clearing from copy code

Buffer must always either contain NULL or valid references
(assuming the dtype supports references) in order to allow
cleanup in case of errors.

It is thus unnecessary to clear buffers before every copy, if
they contain NULL, all is fine. If they contain non-NULL, we should
also DECREF those references (so it would be incorrect as well).

Buffers thus need the memset exactly once: directly upon allcoation
(which we do now).  After this any transfer from/to the buffer needs
to ensure that the buffer is always in a good state.

vorfol committed:
xfail some

Commit:
MAINT: Remove unsafe unions from `np.core.fromnumeric`

Commit:
MAINT: Remove unsafe unions from `np.core.function_base`

Commit:
MAINT: Remove unsafe unions from `np.core.numeric`

Commit:
MAINT: Remove unsafe unions from `np`

Commit:
MAINT: Remove unsafe unions from `np.typing._callable`

vorfol committed:
jobs = 0
use setuptools, wheel without versions

Commit:
TST: Fix a typing failure for python 3.7

Mypy 0.790 does not yet contain annotations for `typing_extensions.SupportsIndex`, so we'll have to define them ourselves

charris committed:
Merge pull request #18887 from BvB93/supports_index

MAINT: Relax certain integer-type constraints

seiko2plus committed:
BLD: remove unnecessary flag `-faltivec` on macOS

  Supporting PowerPC/AltiVec on macOS is no longer necessary,
  even if the Mac/G5 is still running having `-faltivec` or
  replacing it with the new AltiVec flag `-maltivec` wouldn't
  increase that much performance without raw SIMD.

  note: the flag was enabled on non-intel platforms which
  causes fatal build errors on macOS/arm64.

seiko2plus committed:
MAINT: Avoid treating _SIMD module build warnings as errors

dongkeun-oh committed:
BUG FIX for MINGW : threads.h existence test requires GLIBC > 2.12

charris committed:
Merge pull request #18915 from BvB93/unsafe

MAINT: Remove unsafe unions and ABCs from return-annotations

charris committed:
Merge pull request #18921 from charris/backport-18750

MAINT: Allow more recursion depth for scalar tests.

charris committed:
Merge pull request #18923 from charris/backport-18900

BLD: remove unnecessary flag `-faltivec` on macOS

charris committed:
Merge pull request #18925 from charris/backport-18910

BUG: for MINGW,  threads.h existence test requires GLIBC > 2.12

charris committed:
Merge pull request #18922 from charris/backport-18813

BUG: Initialize the full nditer buffer in case of error

charris committed:
Merge pull request #18924 from charris/backport-18903

MAINT, CI: treats _SIMD module build warnings as errors through CI only

charris committed:
BUG: Make changelog recognize ``gh-`` as PR number prefix.

The changelog.py script was only recognizing fast forward merges
where the PR number was of the form "(#xxxxx)", but some people
were merging using the "(gh-xxxxx)" form. Fix that.

charris committed:
DOC: Update 1.20.0-changelog.rst.

charris committed:
DOC: Update 1.19.0-changelog.rst.

charris committed:
DOC: Update 1.18.0-changelog.rst.

charris committed:
DOC: Update 1.17.0-changelog.rst.

charris committed:
Merge pull request #18941 from charris/fix-changelog-script

BUG: Make changelog recognize gh- as a PR number prefix.

charris committed:
REL, DOC: Prepare for the NumPy 1.20.3 release.

* Update 1.20.3-notes.rst
* Create 1.20.3-changelog.rst
* Fix typo in 1.20.2-notes.rst

[skip ci]

charris committed:
Merge pull request #18948 from charris/prepare-1.20.3-release

REL, DOC: Prepare for the NumPy 1.20.3 release.

charris committed:
REL: NumPy 1.20.3 release.

charris committed:
BUG: Fix failing mypy test in 1.20.x.

charris committed:
Merge pull request #18953 from charris/fix-mypy-test

BUG: Fix failing mypy test in 1.20.x.

vorfol committed:
cpp __func__ exclude

vorfol committed:
fix for python 3.11
1. use Py_SET_TYPE
2. _Py_HashDouble is changed

vorfol committed:
Merge tag 'v1.20.3' into vms
NumPy 1.20.3 release

vorfol committed:
Npy_HashDouble

charris commented:
You seem to have pulled in a lot of irrelevant changes. You can try `git rebase --no-ff main` to fix that. In general, rebasing is a cleaner way to go. There are other ways to fix things up if you get into trouble.
